File: BFS/Java/BFSTest.java
package BFS.Java;

/**
 * Test class for the Breadth-First Search implementation.
 * This class provides methods to test various operations of the BFS algorithm.
 */
public class BFSTest {
    
    /**
     * Main method to run the tests.
     * 
     * @param args Command line arguments (not used)
     */
    public static void main(String[] args) {
        System.out.println("Running Breadth-First Search Tests...");
        
        testTraversal();
        testShortestPath();
        testHasPath();
        testConnectedComponents();
        testLevels();
        testBipartite();
        
        System.out.println("All tests completed.");
    }
    
    /**
     * Test the BFS traversal operation.
     */
    private static void testTraversal() {
        System.out.println("\n=== Testing BFS Traversal ===");
        
        // Create a simple graph
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        
        node1.addNeighbor(node2);
        node1.addNeighbor(node3);
        node2.addNeighbor(node4);
        node3.addNeighbor(node5);
        
        BFS bfs = new BFS(node1);
        
        // Test traversal
        System.out.println("BFS Traversal (expected: 1, 2, 3, 4, 5):");
        Node[] traversalResult = bfs.traverse();
        
        if (traversalResult.length == 0) {
            System.out.println("FAIL: Traversal returned empty array");
        } else {
            System.out.print("Traversal result: ");
            for (Node node : traversalResult) {
                System.out.print(node.getValue() + " ");
            }
            System.out.println();
            
            // Check if the first node is the start node
            if (traversalResult[0].getValue() == 1) {
                System.out.println("PASS: First node is the start node (1)");
            } else {
                System.out.println("FAIL: First node is not the start node");
            }
            
            // Check if the second level nodes (2 and 3) come before the third level nodes (4 and 5)
            boolean level2BeforeLevel3 = false;
            for (int i = 0; i < traversalResult.length; i++) {
                if (traversalResult[i].getValue() == 2 || traversalResult[i].getValue() == 3) {
                    for (int j = i + 1; j < traversalResult.length; j++) {
                        if (traversalResult[j].getValue() == 4 || traversalResult[j].getValue() == 5) {
                            level2BeforeLevel3 = true;
                            break;
                        }
                    }
                    break;
                }
            }
            
            if (level2BeforeLevel3) {
                System.out.println("PASS: Level 2 nodes come before level 3 nodes");
            } else {
                System.out.println("FAIL: Level 2 nodes do not come before level 3 nodes");
            }
        }
    }
    
    /**
     * Test the shortest path operation.
     */
    private static void testShortestPath() {
        System.out.println("\n=== Testing Shortest Path ===");
        
        // Create a graph with multiple paths
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        Node node6 = new Node(6);
        
        node1.addNeighbor(node2);
        node1.addNeighbor(node3);
        node2.addNeighbor(node4);
        node3.addNeighbor(node5);
        node4.addNeighbor(node6);
        node5.addNeighbor(node6);
        
        BFS bfs = new BFS(node1);
        
        // Test shortest path to node 6
        System.out.println("Shortest path from 1 to 6 (expected: 1, 2, 4, 6 or 1, 3, 5, 6):");
        Node[] pathResult = bfs.findShortestPath(6);
        
        if (pathResult.length == 0) {
            System.out.println("FAIL: Shortest path returned empty array");
        } else {
            System.out.print("Path result: ");
            for (Node node : pathResult) {
                System.out.print(node.getValue() + " ");
            }
            System.out.println();
            
            // Check if the path starts with the start node and ends with the target node
            if (pathResult[0].getValue() == 1 && pathResult[pathResult.length - 1].getValue() == 6) {
                System.out.println("PASS: Path starts with 1 and ends with 6");
            } else {
                System.out.println("FAIL: Path does not start with 1 or does not end with 6");
            }
            
            // Check if the path length is correct (should be 4)
            if (pathResult.length == 4) {
                System.out.println("PASS: Path length is 4");
            } else {
                System.out.println("FAIL: Path length is not 4");
            }
        }
        
        // Test shortest path to a non-existent node
        System.out.println("\nShortest path from 1 to 7 (expected: empty path):");
        Node[] nonExistentPath = bfs.findShortestPath(7);
        
        if (nonExistentPath.length == 0) {
            System.out.println("PASS: Shortest path to non-existent node returned empty array");
        } else {
            System.out.println("FAIL: Shortest path to non-existent node did not return empty array");
        }
    }
    
    /**
     * Test the hasPath operation.
     */
    private static void testHasPath() {
        System.out.println("\n=== Testing Has Path ===");
        
        // Create a graph
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        
        node1.addNeighbor(node2);
        node1.addNeighbor(node3);
        node2.addNeighbor(node4);
        
        BFS bfs = new BFS(node1);
        
        // Test path to existing nodes
        if (bfs.hasPath(4)) {
            System.out.println("PASS: Found path to node 4");
        } else {
            System.out.println("FAIL: Could not find path to node 4");
        }
        
        // Test path to non-existent node
        if (!bfs.hasPath(5)) {
            System.out.println("PASS: Did not find path to node 5");
        } else {
            System.out.println("FAIL: Found path to node 5, but it should not be reachable");
        }
    }
    
    /**
     * Test the findConnectedComponents operation.
     */
    private static void testConnectedComponents() {
        System.out.println("\n=== Testing Connected Components ===");
        
        // Create a graph with multiple components
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        Node node6 = new Node(6);
        
        node1.addNeighbor(node2);
        node2.addNeighbor(node1);
        
        node3.addNeighbor(node4);
        node4.addNeighbor(node3);
        
        node5.addNeighbor(node6);
        node6.addNeighbor(node5);
        
        Node[] allNodes = {node1, node2, node3, node4, node5, node6};
        
        BFS bfs = new BFS(null); // Start node doesn't matter for this test
        
        // Test finding connected components
        Node[][] components = bfs.findConnectedComponents(allNodes);
        
        if (components.length == 0) {
            System.out.println("FAIL: Connected components returned empty array");
        } else {
            System.out.println("Number of connected components: " + components.length);
            
            // Check if the number of components is correct (should be 3)
            if (components.length == 3) {
                System.out.println("PASS: Found 3 connected components");
            } else {
                System.out.println("FAIL: Did not find 3 connected components");
            }
            
            // Print the components
            for (int i = 0; i < components.length; i++) {
                System.out.print("Component " + (i + 1) + ": ");
                for (Node node : components[i]) {
                    System.out.print(node.getValue() + " ");
                }
                System.out.println();
            }
        }
    }
    
    /**
     * Test the findLevels operation.
     */
    private static void testLevels() {
        System.out.println("\n=== Testing Levels ===");
        
        // Create a graph
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        
        node1.addNeighbor(node2);
        node1.addNeighbor(node3);
        node2.addNeighbor(node4);
        node3.addNeighbor(node5);
        
        BFS bfs = new BFS(node1);
        
        // Test finding levels
        int[] levels = bfs.findLevels();
        
        if (levels.length == 0) {
            System.out.println("FAIL: Levels returned empty array");
        } else {
            System.out.println("Levels array length: " + levels.length);
            
            // Check if the number of levels is correct (should be 5 nodes)
            if (levels.length == 5) {
                System.out.println("PASS: Found levels for 5 nodes");
            } else {
                System.out.println("FAIL: Did not find levels for 5 nodes");
            }
            
            // Print the levels
            System.out.print("Levels: ");
            for (int level : levels) {
                System.out.print(level + " ");
            }
            System.out.println();
        }
    }
    
    /**
     * Test the isBipartite operation.
     */
    private static void testBipartite() {
        System.out.println("\n=== Testing Bipartite ===");
        
        // Create a bipartite graph (e.g., a tree)
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        
        node1.addNeighbor(node2);
        node2.addNeighbor(node1);
        
        node1.addNeighbor(node3);
        node3.addNeighbor(node1);
        
        node2.addNeighbor(node4);
        node4.addNeighbor(node2);
        
        node3.addNeighbor(node5);
        node5.addNeighbor(node3);
        
        Node[] bipartiteNodes = {node1, node2, node3, node4, node5};
        
        BFS bfs = new BFS(null); // Start node doesn't matter for this test
        
        // Test if the graph is bipartite
        if (bfs.isBipartite(bipartiteNodes)) {
            System.out.println("PASS: Correctly identified bipartite graph");
        } else {
            System.out.println("FAIL: Did not correctly identify bipartite graph");
        }
        
        // Create a non-bipartite graph (e.g., a triangle)
        Node nodeA = new Node(1);
        Node nodeB = new Node(2);
        Node nodeC = new Node(3);
        
        nodeA.addNeighbor(nodeB);
        nodeB.addNeighbor(nodeA);
        
        nodeB.addNeighbor(nodeC);
        nodeC.addNeighbor(nodeB);
        
        nodeC.addNeighbor(nodeA);
        nodeA.addNeighbor(nodeC);
        
        Node[] nonBipartiteNodes = {nodeA, nodeB, nodeC};
        
        // Test if the graph is bipartite
        if (!bfs.isBipartite(nonBipartiteNodes)) {
            System.out.println("PASS: Correctly identified non-bipartite graph");
        } else {
            System.out.println("FAIL: Did not correctly identify non-bipartite graph");
        }
    }
}

File: BFS/Python/bfs_test.py
"""
Test class for the BFS implementation.
This class provides methods to test various operations of the BFS algorithm.
"""
from bfs_template import BFSTemplate
from bfs_solution import BFSSolution
from bfs import BFS

# Uncomment the line below to use your implementation
# BFS = BFSTemplate
# Comment the line below to use your implementation
BFS = BFS

def test_bfs_traversal():
    """
    Test the BFS traversal operation.
    """
    print("\n=== Testing BFS Traversal ===")
    bfs = BFS()
    
    # Create a simple graph
    # 1 -- 2 -- 5
    # |    |
    # 3 -- 4
    node1 = bfs.add_node(1)
    node2 = bfs.add_node(2)
    node3 = bfs.add_node(3)
    node4 = bfs.add_node(4)
    node5 = bfs.add_node(5)
    
    bfs.add_edge(node1, node2)
    bfs.add_edge(node1, node3)
    bfs.add_edge(node2, node4)
    bfs.add_edge(node2, node5)
    bfs.add_edge(node3, node4)
    
    # Test BFS traversal starting from node1
    result = bfs.bfs_traversal(node1)
    expected = [1, 2, 3, 4, 5]  # One possible BFS order
    
    # Since BFS can have multiple valid orders depending on how neighbors are processed,
    # we'll check if all nodes are visited and the first node is correct
    if len(result) == len(expected) and result[0] == 1:
        print("PASS: BFS traversal visits all nodes and starts with the correct node")
    else:
        print("FAIL: BFS traversal does not visit all nodes or does not start with the correct node")
        print("Expected:", expected)
        print("Got:", result)
    
    # Reset visited status
    bfs.reset_visited()
    
    # Test BFS traversal starting from node3
    result = bfs.bfs_traversal(node3)
    if len(result) == len(expected) and result[0] == 3:
        print("PASS: BFS traversal from node3 visits all nodes and starts with the correct node")
    else:
        print("FAIL: BFS traversal from node3 does not visit all nodes or does not start with the correct node")
    
    # Test BFS traversal with null start node
    result = bfs.bfs_traversal(None)
    if len(result) == 0:
        print("PASS: BFS traversal with null start node returns empty list")
    else:
        print("FAIL: BFS traversal with null start node does not return empty list")

def test_find_path():
    """
    Test the find path operation.
    """
    print("\n=== Testing Find Path ===")
    bfs = BFS()
    
    # Create a simple graph
    # 1 -- 2 -- 5
    # |    |
    # 3 -- 4
    node1 = bfs.add_node(1)
    node2 = bfs.add_node(2)
    node3 = bfs.add_node(3)
    node4 = bfs.add_node(4)
    node5 = bfs.add_node(5)
    
    bfs.add_edge(node1, node2)
    bfs.add_edge(node1, node3)
    bfs.add_edge(node2, node4)
    bfs.add_edge(node2, node5)
    bfs.add_edge(node3, node4)
    
    # Test finding path from node1 to node5
    path = bfs.find_path(node1, node5)
    if len(path) > 0 and path[0] == 1 and path[-1] == 5:
        print("PASS: Path from node1 to node5 found")
        print("Path:", path)
    else:
        print("FAIL: Path from node1 to node5 not found correctly")
    
    # Test finding path from node3 to node5
    path = bfs.find_path(node3, node5)
    if len(path) > 0 and path[0] == 3 and path[-1] == 5:
        print("PASS: Path from node3 to node5 found")
        print("Path:", path)
    else:
        print("FAIL: Path from node3 to node5 not found correctly")
    
    # Test finding path with disconnected nodes
    node6 = bfs.add_node(6)  # Disconnected node
    path = bfs.find_path(node1, node6)
    if len(path) == 0:
        print("PASS: No path found to disconnected node")
    else:
        print("FAIL: Path found to disconnected node, but should not exist")
    
    # Test finding path with null nodes
    path = bfs.find_path(None, node5)
    if len(path) == 0:
        print("PASS: No path found with null start node")
    else:
        print("FAIL: Path found with null start node, but should not exist")

def test_find_level():
    """
    Test the find level operation.
    """
    print("\n=== Testing Find Level ===")
    bfs = BFS()
    
    # Create a simple graph
    # 1 -- 2 -- 5
    # |    |
    # 3 -- 4
    node1 = bfs.add_node(1)
    node2 = bfs.add_node(2)
    node3 = bfs.add_node(3)
    node4 = bfs.add_node(4)
    node5 = bfs.add_node(5)
    
    bfs.add_edge(node1, node2)
    bfs.add_edge(node1, node3)
    bfs.add_edge(node2, node4)
    bfs.add_edge(node2, node5)
    bfs.add_edge(node3, node4)
    
    # Test finding level of node5 from node1
    level = bfs.find_level(node1, 5)
    if level == 2:
        print("PASS: Level of node5 from node1 is 2")
    else:
        print("FAIL: Level of node5 from node1 is not 2, got", level)
    
    # Test finding level of node4 from node1
    level = bfs.find_level(node1, 4)
    if level == 2:
        print("PASS: Level of node4 from node1 is 2")
    else:
        print("FAIL: Level of node4 from node1 is not 2, got", level)
    
    # Test finding level of node3 from node1
    level = bfs.find_level(node1, 3)
    if level == 1:
        print("PASS: Level of node3 from node1 is 1")
    else:
        print("FAIL: Level of node3 from node1 is not 1, got", level)
    
    # Test finding level of non-existent value
    level = bfs.find_level(node1, 6)
    if level == -1:
        print("PASS: Level of non-existent value is -1")
    else:
        print("FAIL: Level of non-existent value is not -1, got", level)
    
    # Test finding level with null start node
    level = bfs.find_level(None, 5)
    if level == -1:
        print("PASS: Level with null start node is -1")
    else:
        print("FAIL: Level with null start node is not -1, got", level)

def test_connected_components():
    """
    Test the count connected components operation.
    """
    print("\n=== Testing Connected Components ===")
    bfs = BFS()
    
    # Test empty graph
    count = bfs.count_connected_components()
    if count == 0:
        print("PASS: Empty graph has 0 connected components")
    else:
        print("FAIL: Empty graph does not have 0 connected components, got", count)
    
    # Create a graph with two connected components
    # Component 1: 1 -- 2 -- 3
    # Component 2: 4 -- 5
    node1 = bfs.add_node(1)
    node2 = bfs.add_node(2)
    node3 = bfs.add_node(3)
    node4 = bfs.add_node(4)
    node5 = bfs.add_node(5)
    
    bfs.add_edge(node1, node2)
    bfs.add_edge(node2, node3)
    bfs.add_edge(node4, node5)
    
    count = bfs.count_connected_components()
    if count == 2:
        print("PASS: Graph has 2 connected components")
    else:
        print("FAIL: Graph does not have 2 connected components, got", count)
    
    # Add a third component (isolated node)
    node6 = bfs.add_node(6)
    
    count = bfs.count_connected_components()
    if count == 3:
        print("PASS: Graph has 3 connected components after adding isolated node")
    else:
        print("FAIL: Graph does not have 3 connected components after adding isolated node, got", count)
    
    # Connect the components
    bfs.add_edge(node3, node4)
    
    count = bfs.count_connected_components()
    if count == 2:
        print("PASS: Graph has 2 connected components after connecting two components")
    else:
        print("FAIL: Graph does not have 2 connected components after connecting two components, got", count)

def test_bipartite():
    """
    Test the is bipartite operation.
    """
    print("\n=== Testing Bipartite ===")
    bfs = BFS()
    
    # Test empty graph
    if bfs.is_bipartite(None):
        print("PASS: Empty graph is bipartite")
    else:
        print("FAIL: Empty graph is not considered bipartite")
    
    # Create a bipartite graph
    # 1 -- 2
    # |    |
    # 3 -- 4
    node1 = bfs.add_node(1)
    node2 = bfs.add_node(2)
    node3 = bfs.add_node(3)
    node4 = bfs.add_node(4)
    
    bfs.add_edge(node1, node2)
    bfs.add_edge(node2, node3)
    bfs.add_edge(node3, node4)
    bfs.add_edge(node4, node1)
    
    if bfs.is_bipartite(node1):
        print("PASS: Square graph is bipartite")
    else:
        print("FAIL: Square graph is not considered bipartite")
    
    # Create a non-bipartite graph (odd cycle)
    # 1 -- 2
    # |   /
    # 3 -
    bfs.clear()
    node1 = bfs.add_node(1)
    node2 = bfs.add_node(2)
    node3 = bfs.add_node(3)
    
    bfs.add_edge(node1, node2)
    bfs.add_edge(node2, node3)
    bfs.add_edge(node3, node1)
    
    if not bfs.is_bipartite(node1):
        print("PASS: Triangle graph is not bipartite")
    else:
        print("FAIL: Triangle graph is considered bipartite, but should not be")

def main():
    """
    Main method to run the tests.
    """
    print("Running BFS Tests...")
    print("Using implementation:", "BFSSolution" if BFS == BFSSolution else "BFSTemplate")
    
    test_bfs_traversal()
    test_find_path()
    test_find_level()
    test_connected_components()
    test_bipartite()
    
    print("All tests completed.")

if __name__ == "__main__":
    main()
